Enhance canary in telescope function

OBarronCS opened the pull request:
This PR adds text to indicate a canary value whenever it is encountered in the telescope function.

![canary](https://github.com/user-attachments/assets/fa7c71f4-f150-4d4a-b390-2a343dfdcb07)
![reg_canary](https://github.com/user-attachments/assets/831d9cfc-8d76-4a52-bc4a-ab6c1e42bc5a)

OBarronCS committed:
enhance canary

OBarronCS commented on pwndbg/enhance.py:
```diff
@@ -44,6 +44,14 @@ def format_small_int_pair(first: int, second: int) -> Tuple[str, str]:
 
 
 def int_str(value: int) -> str:
```
Do we want to rename this function given it's expanded role?

OBarronCS commented on pwndbg/enhance.py:
```diff
@@ -44,6 +44,14 @@ def format_small_int_pair(first: int, second: int) -> Tuple[str, str]:
+    First, it checks if the integer is a stack canary.
+    Then, it attempts to interpret the bytes as a short strings
+    """
+    if value == pwndbg.commands.canary.canary_value()[0]:
```
The canary_value() function is currently not cached. We should probably cache it, because this get called VERY often, and each time it will read from memory.

OBarronCS replied:
It just needs be cached in a smart way - wait for the memory page where the canary is to be loaded in the first place, then assume it won't change during the program execution?

disconnect3d replied:
The problem is: it may change ;D.

disconnect3d replied:
Btw this calls a function from `pwndbg.commands.*` - those should likely almost never be called externally... We should probably move that canary function to some lib counterpart.

disconnect3d replied:
This should also be gated somehow so that this is never called when:
1) debugging linux kernel
2) debugging embedded devices

etc.

OBarronCS replied:
:0 didn't know the stack canary could change during the invocation of the program. Does the value pointed at by AT_RANDOM change? Or is there another way it changes?

disconnect3d commented:
Wait, why is the RAX on the screenshot not in hex but in decimal? :D

OBarronCS commented:
Oop - had too much tunnel vision on the `(canary)` hahaha

OBarronCS committed:
hexify

OBarronCS commented:
Fixed: ![image](https://github.com/user-attachments/assets/dbc86d6a-c196-4eca-b651-5414a69ad04c)

disconnect3d commented:
Few things here:

1. We call a function from `pwndbg.commands.*` - we should probably move that canary function to some lib counterpart so that we don't require this part of code to import from a command
2. This needs to be gated somehow: we don't want to call this code ever (and ideally, don't even perform any check) when we debug 1) linux kernel; 2) embedded device; 3) any other target where reading memory is not cheap, heh.

OBarronCS commented:
Given the performance concerns, I'm closing this for now. If anyone thinks of a clean way to implement this that addresses the comments above, please feel free to take this up.
